README.md
# TinyBit Core #
**Contributors:** [(this-should-be-a-list-of-wordpress.org-userid's)](https://profiles.wordpress.org/(this-should-be-a-list-of-wordpress.org-userid's)/)  
**Tags:** comments, spam  
**Requires at least:** 4.5  
**Tested up to:** 5.8.2  
**Requires PHP:** 5.6  
**Stable tag:** 0.1.0  
**License:** GPLv2 or later  
**License URI:** https://www.gnu.org/licenses/gpl-2.0.html  

Here is a short description of the plugin.  This should be no more than 150 characters.  No markup here.

## Description ##

This is the long description.  No limit, and you can use Markdown (as well as in the following sections).

For backwards compatibility, if this section is missing, the full length of the short description will be used, and
Markdown parsed.

A few notes about the sections above:

*   "Contributors" is a comma separated list of wp.org/wp-plugins.org usernames
*   "Tags" is a comma separated list of tags that apply to the plugin
*   "Requires at least" is the lowest version that the plugin will work on
*   "Tested up to" is the highest version that you've *successfully used to test the plugin*. Note that it might work on
higher versions... this is just the highest one you've verified.
*   Stable tag should indicate the Subversion "tag" of the latest stable version, or "trunk," if you use `/trunk/` for
stable.

    Note that the `readme.txt` of the stable tag is the one that is considered the defining one for the plugin, so
if the `/trunk/readme.txt` file says that the stable tag is `4.3`, then it is `/tags/4.3/readme.txt` that'll be used
for displaying information about the plugin.  In this situation, the only thing considered from the trunk `readme.txt`
is the stable tag pointer.  Thus, if you develop in trunk, you can update the trunk `readme.txt` to reflect changes in
your in-development version, without having that information incorrectly disclosed about the current stable version
that lacks those changes -- as long as the trunk's `readme.txt` points to the correct stable tag.

    If no stable tag is provided, it is assumed that trunk is stable, but you should specify "trunk" if that's where
you put the stable version, in order to eliminate any doubt.

## Installation ##

This section describes how to install the plugin and get it working.

e.g.

1. Upload `plugin-name.php` to the `/wp-content/plugins/` directory
1. Activate the plugin through the 'Plugins' menu in WordPress
1. Place `<?php do_action('plugin_name_hook'); ?>` in your templates

## Frequently Asked Questions ##

### A question that someone might have ###

An answer to that question.

### What about foo bar? ###

Answer to foo bar dilemma.

## Screenshots ##

1. This screen shot description corresponds to screenshot-1.(png|jpg|jpeg|gif). Note that the screenshot is taken from
the /assets directory or the directory that contains the stable readme.txt (tags or trunk). Screenshots in the /assets
directory take precedence. For example, `/assets/screenshot-1.png` would win over `/tags/4.3/screenshot-1.png`
(or jpg, jpeg, gif).
2. This is the second screen shot

## Changelog ##

### 1.0 ###
* A change since the previous version.
* Another change.

### 0.5 ###
* List versions from most recent at top to oldest at bottom.

## Upgrade Notice ##

### 1.0 ###
Upgrade notices describe the reason a user should upgrade.  No more than 300 characters.

### 0.5 ###
This version fixes a security related bug.  Upgrade immediately.

## Arbitrary section ##

You may provide arbitrary sections, in the same format as the ones above.  This may be of use for extremely complicated
plugins where more information needs to be conveyed that doesn't fit into the categories of "description" or
"installation."  Arbitrary sections will be shown below the built-in sections outlined above.

## A brief Markdown Example ##

Ordered list:

1. Some feature
1. Another feature
1. Something else about the plugin

Unordered list:

* something
* something else
* third thing

Here's a link to [WordPress](https://wordpress.org/ "Your favorite software") and one to [Markdown's Syntax Documentation][markdown syntax].
Titles are optional, naturally.

[markdown syntax]: https://daringfireball.net/projects/markdown/syntax
            "Markdown is what the parser uses to process much of the readme file"

Markdown uses email style notation for blockquotes and I've been told:
> Asterisks for *emphasis*. Double it up  for **strong**.

`<?php code(); // goes in backticks ?>`


CHANGELOG.md
# Changelog

## [0.3.0] - 2022-03-16

### Added
* Adds a WP-CLI command for auditing `<head>` meta values [[#6](https://github.com/wearetinybit/tinybit-core-plugin/pull/6)].

## [0.2.3] - 2022-02-21

### Added
* Disables WordPress core's enhanced lazy-loading [[#4](https://github.com/wearetinybit/tinybit-core-plugin/pull/4)].

## [0.2.2] - 2022-02-18

### Fixed
* Fixes issue where content would end up repeating [[#3](https://github.com/wearetinybit/tinybit-core-plugin/pull/3)].

## [0.2.1] - 2022-02-18

### Added
* Eager loads images within the first four paragraphs of content [[#2](https://github.com/wearetinybit/tinybit-core-plugin/pull/2)].


tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package TinyBit_Critical_Css
 */

$_tests_dir = getenv( 'WP_TESTS_DIR' );

if ( ! $_tests_dir ) {
	$_tests_dir = rtrim( sys_get_temp_dir(), '/\\' ) . '/wordpress-tests-lib';
}

define( 'WP_TESTS_PHPUNIT_POLYFILLS_PATH', __DIR__ . '/../vendor/yoast/phpunit-polyfills/phpunitpolyfills-autoload.php' );

if ( ! file_exists( $_tests_dir . '/includes/functions.php' ) ) {
	throw new Exception( "Could not find $_tests_dir/includes/functions.php, have you run bin/install-wp-tests.sh ?" );
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/tinybit-core.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/includes/bootstrap.php';

echo PHP_EOL;
echo 'WP version: ' . $GLOBALS['wp_version'];
echo PHP_EOL . PHP_EOL;


tests/test-utils.php
<?php
/**
 * Tests the TBC\Utils class to verify everything works.
 *
 * @package TBC
 */

namespace TBC\Tests;

use TBC\Utils;

/**
 * Tests the TBC\Utils class to verify everything works.
 *
 * @coversDefaultClass TBC\Utils
 */
class Test_Utils extends \WP_UnitTestcase {

	/**
	 * Ensures force_element_attribute works as expected.
	 */
	public function test_force_element_attribute() {
		// Attribute doesn't yet exist.
		$this->assertEquals(
			'<p><img loading="eager" src="https://curbly.test/wp-content/uploads/2020/12/cash-tree-feature-3476-edit_web.jpg" alt="Cash forest - the perfect holiday gift!" width="1000" height="750" class="align size-full wp-image-13791"></p>',
			Utils::force_element_attribute(
				'<p><img src="https://curbly.test/wp-content/uploads/2020/12/cash-tree-feature-3476-edit_web.jpg" alt="Cash forest - the perfect holiday gift!" width="1000" height="750" class="align size-full wp-image-13791"></p>',
				'img',
				'loading',
				'eager'
			)
		);
		$this->assertEquals(
			'<p><img src="https://curbly.test/wp-content/uploads/2020/12/cash-tree-feature-3476-edit_web.jpg" loading="eager" alt="Cash forest - the perfect holiday gift!" width="1000" height="750" class="align size-full wp-image-13791"></p>',
			Utils::force_element_attribute(
				'<p><img src="https://curbly.test/wp-content/uploads/2020/12/cash-tree-feature-3476-edit_web.jpg" loading="lazy" alt="Cash forest - the perfect holiday gift!" width="1000" height="750" class="align size-full wp-image-13791"></p>',
				'img',
				'loading',
				'eager'
			)
		);
	}
}


tests/test-frontend.php
<?php
/**
 * Tests frontend pieces to ensure they work as expected.
 *
 * @package TBC
 */

namespace TBC\Tests;

/**
 * Tests frontend pieces to ensure they work as expected.
 */
class Test_Frontend extends \WP_UnitTestCase {

	/**
	 * Ensure early content filters work as expected.
	 */
	public function test_filter_the_content_early() {
		$original = <<<EOT
Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum. Integer posuere erat a ante venenatis dapibus posuere velit aliquet. Etiam porta sem malesuada magna mollis euismod.

<img src="https://example.com" height="240" width="300">

Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum. Integer posuere erat a ante venenatis dapibus posuere velit aliquet. Etiam porta sem malesuada magna mollis euismod.

Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum. Integer posuere erat a ante venenatis dapibus posuere velit aliquet. Etiam porta sem malesuada magna mollis euismod.
EOT;
		$expected = <<<EOT
<p>Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum. Integer posuere erat a ante venenatis dapibus posuere velit aliquet. Etiam porta sem malesuada magna mollis euismod.</p>
<p><img loading="eager" src="https://example.com" height="240" width="300"></p>
<p>Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum. Integer posuere erat a ante venenatis dapibus posuere velit aliquet. Etiam porta sem malesuada magna mollis euismod.</p>
<p>Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum. Integer posuere erat a ante venenatis dapibus posuere velit aliquet. Etiam porta sem malesuada magna mollis euismod.</p>
EOT;
		$post_id  = $this->factory->post->create(
			[
				'post_content' => $original,
			]
		);
		$this->go_to( get_permalink( $post_id ) );
		$this->assertEquals(
			trim( $expected ),
			trim( apply_filters( 'the_content', $original ) )
		);
	}

}


functions.php
<?php
/**
 * Defines functions for us to use.
 *
 * @package TBC
 */

/**
 * Registers an array of hooks for a class.
 *
 * @param string $class Class name.
 * @param array  $hooks Hooks to register.
 */
function tbc_register_class_hooks( $class, $hooks ) {
	foreach ( tbc_parse_register_class_hooks( $class, $hooks ) as $parsed_hook ) {
		list( $type, $hook, $class, $method, $priority, $args ) = $parsed_hook;
		$function = 'action' === $type ? 'add_action' : 'add_filter';
		$function( $hook, [ $class, $method ], $priority, $args );
	}
}

/**
 * Parses an array of hooks into their appropriate definitions.
 *
 * @param string $class Class name.
 * @param array  $hooks Hooks to register.
 * @return array
 */
function tbc_parse_register_class_hooks( $class, $hooks ) {
	$parsed_hooks = [];
	foreach ( $hooks as $hook_data ) {
		$hook     = array_shift( $hook_data );
		$method   = null;
		$priority = 10;
		$args     = 1;
		if ( isset( $hook_data[0] ) && is_string( $hook_data[0] ) ) {
			$method = array_shift( $hook_data );
		}
		if ( isset( $hook_data[0] ) ) {
			$priority = array_shift( $hook_data );
			if ( isset( $hook_data[0] ) ) {
				$args = array_shift( $hook_data );
			}
		}
		if ( $method ) {
			$type = null;
			if ( 0 === stripos( $method, 'filter_' ) ) {
				$type = 'filter';
			} elseif ( 0 === stripos( $method, 'action_' ) ) {
				$type = 'action';
			}
			$parsed_hooks[] = [
				$type,
				$hook,
				$class,
				$method,
				$priority,
				$args,
			];
		} else {
			$type = null;
			if ( method_exists( $class, 'filter_' . $hook ) ) {
				$type   = 'filter';
				$method = 'filter_' . $hook;
			} elseif ( method_exists( $class, 'action_' . $hook ) ) {
				$type   = 'action';
				$method = 'action_' . $hook;
			}
			$parsed_hooks[] = [
				$type,
				$hook,
				$class,
				$method,
				$priority,
				$args,
			];
		}
	}
	return $parsed_hooks;
}


000-loader.php
<?php
/**
 * Loads what we needs.
 *
 * @package TBC
 */

/**
 * Registers an array of hooks for a class.
 *
 * @param string $class Class name.
 * @param array  $hooks Hooks to register.
 */
function tbc_register_class_hooks( $class, $hooks ) {
	foreach ( tbc_parse_register_class_hooks( $class, $hooks ) as $parsed_hook ) {
		list( $type, $hook, $class, $method, $priority, $args ) = $parsed_hook;
		$function = 'action' === $type ? 'add_action' : 'add_filter';
		$function( $hook, [ $class, $method ], $priority, $args );
	}
}

/**
 * Parses an array of hooks into their appropriate definitions.
 *
 * @param string $class Class name.
 * @param array  $hooks Hooks to register.
 * @return array
 */
function tbc_parse_register_class_hooks( $class, $hooks ) {
	$parsed_hooks = [];
	foreach ( $hooks as $hook_data ) {
		$hook     = array_shift( $hook_data );
		$method   = null;
		$priority = 10;
		$args     = 1;
		if ( isset( $hook_data[0] ) && is_string( $hook_data[0] ) ) {
			$method = array_shift( $hook_data );
		}
		if ( isset( $hook_data[0] ) ) {
			$priority = array_shift( $hook_data );
			if ( isset( $hook_data[0] ) ) {
				$args = array_shift( $hook_data );
			}
		}
		if ( $method ) {
			$type = null;
			if ( 0 === stripos( $method, 'filter_' ) ) {
				$type = 'filter';
			} elseif ( 0 === stripos( $method, 'action_' ) ) {
				$type = 'action';
			}
			$parsed_hooks[] = [
				$type,
				$hook,
				$class,
				$method,
				$priority,
				$args,
			];
		} else {
			$type = null;
			if ( method_exists( $class, 'filter_' . $hook ) ) {
				$type   = 'filter';
				$method = 'filter_' . $hook;
			} elseif ( method_exists( $class, 'action_' . $hook ) ) {
				$type   = 'action';
				$method = 'action_' . $hook;
			}
			$parsed_hooks[] = [
				$type,
				$hook,
				$class,
				$method,
				$priority,
				$args,
			];
		}
	}
	return $parsed_hooks;
}

/**
 * Register the class autoloader
 */
spl_autoload_register(
	function( $class ) {
		$class = ltrim( $class, '\\' );
		if ( 0 !== stripos( $class, 'TBC\\' ) ) {
			return;
		}

		$parts = explode( '\\', $class );
		array_shift( $parts ); // Don't need "TBC".
		$last    = array_pop( $parts ); // File should be 'class-[...].php'.
		$last    = 'class-' . $last . '.php';
		$parts[] = $last;
		$file    = dirname( __FILE__ ) . '/inc/' . str_replace( '_', '-', strtolower( implode( '/', $parts ) ) );
		if ( file_exists( $file ) ) {
			require $file;
		}

	}
);


tinybit-core.php
<?php
/**
 * Plugin Name:     TinyBit Core
 * Plugin URI:      PLUGIN SITE HERE
 * Description:     PLUGIN DESCRIPTION HERE
 * Author:          YOUR NAME HERE
 * Author URI:      YOUR SITE HERE
 * Text Domain:     tinybit-core
 * Domain Path:     /languages
 * Version:         0.3.0
 *
 * @package         TBC
 */

require __DIR__ . '/000-loader.php';

/*
 * TBC\Frontend
 */
tbc_register_class_hooks(
	'TBC\Frontend',
	[
		[ 'the_content', 'filter_the_content_early', 1 ],
	]
);

/*
 * TBC\Media
 */
tbc_register_class_hooks(
	'TBC\Media',
	[
		[ 'wp_generate_attachment_metadata', 10, 2 ],
	]
);

add_filter( 'wp_omit_loading_attr_threshold', '__return_zero' );

/*
 * TBC\Integrations\Cloudflare
 */
tbc_register_class_hooks(
	'TBC\Integrations\Cloudflare',
	[
		[ 'cloudflare_purge_by_url' ],
	]
);

if ( defined( 'WP_CLI' ) && WP_CLI ) {
	WP_CLI::add_command( 'tbc', 'TBC\CLI' );
}


inc/class-media.php
<?php
/**
 * Manages media library functionality.
 *
 * @package TBC
 */

namespace TBC;

/**
 * Manages media library functionality.
 */
class Media {

	/**
	 * Modifies the quality of original images during the upload process.
	 *
	 * @param array   $metadata      Existing attachment metadata.
	 * @param integer $attachment_id ID for the attachment.
	 * @return array
	 */
	public static function filter_wp_generate_attachment_metadata( $metadata, $attachment_id ) {
		self::compress_attachment_original_image( $attachment_id );
		return $metadata;
	}

	/**
	 * Rescales the full-size image for a given attachment ID.
	 *
	 * @param integer $attachment_id ID for the attachment.
	 * @param integer $quality       Compression quality to use.
	 */
	public static function compress_attachment_original_image( $attachment_id, $quality = 70 ) {
		$file = get_attached_file( $attachment_id );
		$type = get_post_mime_type( $attachment_id );

		if ( ! in_array( $type, [ 'image/jpg', 'image/jpeg' ], true ) ) {
			return;
		}

		$gd_filter = function() {
			return array( 'WP_Image_Editor_GD' );
		};

		add_filter(
			'wp_image_editors',
			$gd_filter
		);

		$editor = wp_get_image_editor( $file );
		if ( ! is_wp_error( $editor ) ) {
			$result = $editor->set_quality( $quality );
			if ( ! is_wp_error( $result ) ) {
				$editor->save( $file );
			}
		}

		remove_filter(
			'wp_image_editors',
			$gd_filter
		);
	}

}


inc/class-utils.php
<?php
/**
 * Various utility functions.
 *
 * @package TBC
 */

namespace TBC;

/**
 * Various utility functions.
 */
class Utils {

	/**
	 * Forces a specific value for a specific HTML element attribute.
	 *
	 * @param string $content   Content to process.
	 * @param string $element   Element to inspect.
	 * @param string $attribute Attribute to enforce.
	 * @param string $value     Attribute value to enforce.
	 * @return content
	 */
	public static function force_element_attribute( $content, $element, $attribute, $value ) {
		$content = preg_replace_callback(
			'#<' . $element . '([^>]*)>#',
			function( $match ) use ( $element, $attribute, $value ) {
				$full = $match[0];
				$full = preg_replace(
					'#(<' . $element . '[^>]+' . $attribute . '=["\'])([^"\']+)#',
					'$1' . $value,
					$full,
					-1,
					$count
				);
				if ( ! $count ) {
					$full = str_replace(
						'<' . $element . ' ',
						'<' . $element . ' ' . $attribute . '="' . $value . '" ',
						$full
					);
				}
				return $full;
			},
			$content
		);
		return $content;
	}
}


inc/class-cli.php
<?php
/**
 * TinyBit Core Utilities
 *
 * @package TBC
 */

namespace TBC;

use WP_CLI;
use WP_Error;
use WP_Query;

/**
 * TinyBit Core Utilities.
 */
class CLI {

	/**
	 * Whether or not the header has been rendered.
	 *
	 * @var boolean
	 */
	private static $rendered_header;

	/**
	 * Whether or not the footer has been rendered.
	 *
	 * @var boolean
	 */
	private static $rendered_footer;

	/**
	 * Generates width-constrained images at the same proportion as the original image.
	 *
	 * ## OPTIONS
	 *
	 * <file>
	 * : Original image file.
	 *
	 * --widths=<widths>
	 * : Comma-separated image widths to generate.
	 *
	 * [--quality=<quality>]
	 * : Compression quality.
	 * ---
	 * default: 70
	 * ---
	 *
	 * [--compress]
	 * : Run the image through additional compression.
	 *
	 * @subcommand generate-image-widths
	 */
	public function generate_image_widths( $args, $assoc_args ) {

		list( $file ) = $args;
		if ( ! file_exists( $file ) ) {
			WP_CLI::error( 'File does not exist.' );
		}

		$widths = array_unique( array_map( 'intval', explode( ',', $assoc_args['widths'] ) ) );
		sort( $widths );
		$editor = wp_get_image_editor( $file );
		if ( is_wp_error( $editor ) ) {
			WP_CLI::error( $editor );
		}
		$size = $editor->get_size();
		foreach ( $widths as $width ) {
			if ( $width >= $size['width'] ) {
				WP_CLI::warning( sprintf( 'Skipping: provided width \'%d\' exceeds original image width \'%d\'.', $width, $size['width'] ) );
				continue;
			}
			$editor = wp_get_image_editor( $file );
			if ( is_wp_error( $editor ) ) {
				WP_CLI::error( $editor );
			}
			$editor->resize( $width, null, false );
			$editor->set_quality( $assoc_args['quality'] );
			$dir       = dirname( $file );
			$ext       = strtolower( pathinfo( $file, PATHINFO_EXTENSION ) );
			$new_file  = wp_basename( $file, ".$ext" );
			$new_size  = $editor->get_size();
			$new_file  = preg_replace(
				'#(-[\d]+x[\d]+)?$#',
				'',
				$new_file
			);
			$new_file  = preg_replace(
				'#(\@[\d]x?)?$#',
				'',
				$new_file
			);
			$new_file .= '-' . (int) $new_size['width'] . 'x' . (int) $new_size['height'];
			$editor->save( trailingslashit( $dir ) . $new_file, mime_content_type( $file ) );
			$full_file = trailingslashit( $dir ) . $new_file . '.' . $ext;
			if ( ! empty( $assoc_args['compress'] ) ) {
				$orig_kb = round( ( filesize( $full_file ) / 1000 ), 2 );
				$ret     = self::compress_image_inline_with_tinypng( $full_file );
				if ( is_wp_error( $ret ) ) {
					WP_CLI::error( $ret );
				}
				clearstatcache( false, $full_file );
				$new_kb = round( ( filesize( $full_file ) / 1000 ), 2 );
				WP_CLI::log( sprintf( 'Generated %s (%dkb->%dkb)', $new_file . '.' . $ext, $orig_kb, $new_kb ) );
			} else {
				$new_kb = round( ( filesize( $full_file ) / 1000 ), 2 );
				WP_CLI::log( sprintf( 'Generated %s (%dkb)', $new_file . '.' . $ext, $new_kb ) );
			}
		}

		WP_CLI::success( 'Image widths created.' );
	}

	/**
	 * Compresses images inline using TinyPNG.
	 *
	 * ## OPTIONS
	 *
	 * <files>...
	 * : Original image files.
	 *
	 * @subcommand compress-images
	 */
	public function compress_images( $args ) {
		foreach ( $args as $file ) {
			if ( ! file_exists( $file ) ) {
				WP_CLI::warning( sprintf( 'Skipping missing file: %s', $file ) );
				continue;
			}
			$orig_kb = round( ( filesize( $file ) / 1000 ), 2 );
			$ret     = self::compress_image_inline_with_tinypng( $file );
			if ( is_wp_error( $ret ) ) {
				WP_CLI::warning( $ret->get_error_message() );
				continue;
			}
			clearstatcache( false, $file );
			$new_kb = round( ( filesize( $file ) / 1000 ), 2 );
			WP_CLI::log(
				sprintf(
					'Compressed %s from %dkb to %dkb',
					wp_basename( $file ),
					$orig_kb,
					$new_kb
				)
			);
		}
		WP_CLI::success( 'Images compressed.' );
	}

	/**
	 * Audit posts for their og:image and schema primary image.
	 *
	 * ## OPTIONS
	 *
	 * <sitemap>
	 * : Sitemap to pull URLs from.
	 *
	 * [--format=<format>]
	 * : Output format for the results.
	 * ---
	 * default: table
	 * options:
	 *   - table
	 *   - csv
	 *   - json
	 *   - html
	 * ---
	 *
	 * @subcommand audit-head-meta
	 */
	public function audit_head_meta( $args, $assoc_args ) {
		$results = [];

		list( $sitemap ) = $args;

		$response = \WP_CLI\Utils\http_request( 'GET', $sitemap );
		if ( 200 !== $response->status_code ) {
			WP_CLI::error( 'Unable to fetch sitemap' );
		}

		preg_match_all( '#<loc>(.+)</loc>#Us', $response->body, $matches );
		$urls = ! empty( $matches[1] ) ? $matches[1] : [];

		foreach ( $urls as $url ) {
			$response = \WP_CLI\Utils\http_request( 'GET', $url );
			$output   = $response->body;

			$og_image = '';
			if ( preg_match( '#<meta property="og:image" content="([^"]+)"#', $output, $matches ) ) {
				$og_image = $matches[1];
			}
			$schema_image = '';
			if ( preg_match(
				'#<script type="application/ld\+json" class="yoast-schema-graph">(.+)</script>#',
				$output,
				$matches
			) ) {
				$schema = json_decode( $matches[1], true );
				if ( ! empty( $schema['@graph'] ) ) {
					foreach ( $schema['@graph'] as $piece ) {
						if ( 'ImageObject' === $piece['@type'] ) {
							$schema_image = $piece['url'];
							break;
						}
					}
				}
			}

			$results[] = [
				'url'          => $url,
				'og_image'     => $og_image,
				'schema_image' => $schema_image,
			];
		}

		$headers = [];
		if ( ! empty( $results ) ) {
			$headers = array_keys( $results[0] );
		}

		if ( 'html' === $assoc_args['format'] ) {
			$output = '<table><thead><tr>';
			foreach ( $headers as $header ) {
				$output .= '<th>' . $header . '</th>';
			}
			$output .= '</tr></thead>';
			$output .= '<tbody>';
			foreach ( $results as $result ) {
				$output .= '<tr>';
				foreach ( $result as $key => $value ) {
					if ( false !== stripos( $key, '_image' ) && ! empty( $value ) ) {
						$output .= '<td><img loading="lazy" width="300" src="' . $value . '"></td>';
					} elseif ( false !== stripos( $key, 'url' ) && ! empty( $value ) ) {
						$output .= '<td><a target="_blank" href="' . $value . '">' . $value . ' </a></td>';
					} else {
						$output .= '<td>' . $value . '</td>';
					}
				}
				$output .= '</tr>';
			}
			$output .= '</tbody></table>';
			echo $output;
		} else {
			WP_CLI\Utils\format_items( $assoc_args['format'], $results, $headers );
		}
	}

	/**
	 * Compresses an image in-place using TinyPNG.
	 *
	 * @param string $file Image file path.
	 * @return true|WP_Error
	 */
	private static function compress_image_inline_with_tinypng( $file ) {

		if ( ! defined( 'TBC_TINYPNG_API_KEY' ) ) {
			return new WP_Error( 'missing-api-key', 'Constant TBC_TINYPNG_API_KEY is not defined.' );
		}

		$response = wp_remote_post(
			'https://api.tinify.com/shrink',
			[
				'headers' => [
					'Authorization' => 'Basic ' . base64_encode( 'api:' . TBC_TINYPNG_API_KEY ),
					'Content-Type'  => 'multipart/form-data',
				],
				'body'    => file_get_contents( $file ),
			]
		);
		if ( is_wp_error( $response ) ) {
			return $response;
		}
		$status = wp_remote_retrieve_response_code( $response );
		$body   = wp_remote_retrieve_body( $response );
		if ( 200 === $status || 201 === $status ) {
			$body           = json_decode( $body, true );
			$image_response = wp_remote_get( $body['output']['url'] );
			if ( is_wp_error( $image_response ) ) {
				return $image_response;
			}
			$image_status = wp_remote_retrieve_response_code( $response );
			if ( 200 === $image_status || 201 === $image_status ) {
				file_put_contents( $file, wp_remote_retrieve_body( $image_response ) );
			} else {
				return new WP_Error(
					'download-http-error',
					sprintf( 'Unexpected error downloading compressed image (HTTP %d)', $image_status )
				);
			}
		} else {
			return new WP_Error(
				'compress-http-error',
				sprintf( 'Unexpected error compressing image (HTTP %d)', $status )
			);
		}

		return true;
	}
}
